lib/Parse/ParseDecl.cpp

```diff
@@ -3618,19 +3618,34 @@ void Parser::ParseStructDeclaration(
     } else
       DeclaratorInfo.D.SetIdentifier(nullptr, Tok.getLocation());
 
+    CachedTokens *BoundsExprTokens = nullptr;
     if (TryConsumeToken(tok::colon)) {
-      ExprResult Res(ParseConstantExpression());
-      if (Res.isInvalid())
-        SkipUntil(tok::semi, StopBeforeMatch);
-      else
-        DeclaratorInfo.BitfieldSize = Res.get();
+      bool IsBoundsExpr = false;
+      if (getLangOpts().CheckedC && Tok.getKind() == tok::identifier) {
+        IdentifierInfo *Ident = Tok.getIdentifierInfo();
+        IsBoundsExpr = Ident == Ident_byte_count || Ident == Ident_count ||
+                       Ident == Ident_bounds;
+      }
```

Comment:
This logic--peeking at the upcoming tokens and seeing if they form the start of a valid bounds expression--is now included inline in a few places in clang.  Would it make sense to consolidate them into a shared helper function, ideally close to the code that parses bounds expressions?  That might help to consolidate all logic that understands how to detect and parse them and increase the likelihood that it's kept consistent.